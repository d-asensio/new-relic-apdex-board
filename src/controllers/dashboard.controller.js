
/**
 * Mediates between the user input, the data and the view.
 */
export class DashboardController {
  constructor (service, view) {
    this._isInitialized = false

    this._service = service
    this._view = view

    this._attachViewListeners()
  }

  async init () {
    await this._service.init()
    this._isInitialized = true
  }

  isInitialized () {
    return this._isInitialized
  }

  renderToDom (element) {
    this._isInitializedOrThrow()
    this._isValidElementOrThrow(element)

    const viewElement = this._getViewElement()
    element.append(viewElement)
  }

  _getViewElement () {
    return this._view.create(
      this._service.dashboard
    )
  }

  _attachViewListeners () {
    this._view.onClickApp(
      this._handleClickApp.bind(this)
    )

    this._view.onToggleLayout(
      this._handleLayoutToggle.bind(this)
    )
  }

  _handleClickApp ({ appId }) {
    const clickedApp = this._service.getAppById(appId)

    alert(`Release number: ${clickedApp.version}`)
  }

  _handleLayoutToggle ({ isActive }) {
    if (isActive) {
      this._view.setLayoutToList()
    } else {
      this._view.setLayoutToGrid()
    }
  }

  _isValidElementOrThrow (element) {
    if (!(element instanceof HTMLElement)) {
      throw new TypeError(
        'The provided element is not valid.'
      )
    }
  }

  _isInitializedOrThrow () {
    if (!this.isInitialized()) {
      throw new Error(
        'The controller is not yet initialized. You should call to the "init ()" asynchronous method and await for it' +
        'to finis before proceeding.'
      )
    }
  }
}
