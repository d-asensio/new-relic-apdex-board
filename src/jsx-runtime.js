export default function createElement (tag, props, ...children) {
  const element = document.createElement(tag)

  if (props !== null) {
    _addPropsToElement(props, element)
  }

  _appendChildrenToElement(children, element)

  return element
}

function _addPropsToElement (props, element) {
  const { className, ...attributes } = props

  _addAttributesToElement(attributes, element)

  if (className) {
    _addClassNameToElement(className, element)
  }
}

function _addAttributesToElement (attributes, element) {
  const attributeEntries = Object.entries(attributes)

  for (const [attrName, attrValue] of attributeEntries) {
    _attributeSupportedByElementOrThrow(attrName, element)

    element.setAttribute(attrName, attrValue)
  }
}

function _attributeSupportedByElementOrThrow (attributeName, element) {
  if (!_isAttributeSupportedByElement(attributeName, element)) {
    throw new Error(
      `A "${element.tagName.toLowerCase()}" element do not support the attribute "${attributeName}".`
    )
  }
}

function _isAttributeSupportedByElement (attributeName, element) {
  if (_isDataAttribute(attributeName)) return true
  if (_isAriaAttribute(attributeName)) return true

  return attributeName in element
}

function _isDataAttribute (attributeName) {
  return attributeName.startsWith('data-')
}

function _isAriaAttribute (attributeName) {
  return attributeName.startsWith('aria-')
}

function _addClassNameToElement (className, element) {
  const sanitizedClassNames = _sanitizeClassName(className)

  for (const sanitizedClass of sanitizedClassNames) {
    element.classList.add(sanitizedClass)
  }
}

function _sanitizeClassName (className) {
  return className
    .split(' ')
    .filter(Boolean)
}

function _appendChildrenToElement (children, element) {
  for (const child of children) {
    if (Array.isArray(child)) {
      _appendChildrenToElement(child, element)
    } else {
      _appendChildToElement(child, element)
    }
  }
}

function _appendChildToElement (child, element) {
  const safeChildNode = _createSafeChildNode(child, element)

  element.appendChild(safeChildNode)
}

function _createSafeChildNode (child) {
  if (child instanceof HTMLElement || child instanceof Text) {
    return child
  }

  return document.createTextNode(child)
}
