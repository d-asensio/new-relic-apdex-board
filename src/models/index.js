export * from './application.model.js'
export * from './dashboard.model.js'
export * from './host.model.js'
