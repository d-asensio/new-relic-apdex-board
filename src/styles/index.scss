@import "reset";

@import "tokens/all";
@import "utilities/all";

@import "views/all";

body {
  background: #f5f5f5;
  
  @include sp-inset($space--large);
}
