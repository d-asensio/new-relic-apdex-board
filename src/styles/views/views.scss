@import "dashboard";
@import "host";
@import "toggle";