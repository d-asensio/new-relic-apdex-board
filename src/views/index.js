export * from './dashboard.view.js'
export * from './host.view.js'
export * from './toggle.view.js'
