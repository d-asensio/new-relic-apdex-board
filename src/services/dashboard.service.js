import { Dashboard, Application, Host } from '../models'

const API_URL = 'data/host-app-data.json'

/**
 * Manages the data of the dashboard, meaning that it kwows how to create models and how to get data to create them.
 */
export class DashboardService {
  constructor () {
    this._dashboard = new Dashboard({
      user: 'this.email.is.so.long@newrelic.com'
    })
  }

  get dashboard () { return this._dashboard }

  async init () {
    const appRecords = await this._fetchAppRecordsFromApi()

    const sortedRecords = this._sortAppRecordsByApdex(appRecords)
    this._createAppModelsFromRecords(sortedRecords)
  }

  addAppToHosts (app, hosts) {
    this._appExistsOrThrow(app)

    for (const hostId of hosts) {
      const host = this._getOrCreateHostById(hostId)
      host.addApp(app)
    }
  }

  removeAppFromHosts (app, hosts) {
    this._appExistsOrThrow(app)

    for (const hostId of hosts) {
      this._hostIdExistsOrThrow(hostId)

      const host = this._getHostById(hostId)
      host.removeApp(app)
    }
  }

  getTopAppsByHost (hostId) {
    this._hostIdExistsOrThrow(hostId)

    const host = this._getHostById(hostId)

    return host.getTopApps(25)
  }

  getAppById (appId) {
    return this._dashboard.getAppById(appId)
  }

  async _fetchAppRecordsFromApi () {
    const res = await fetch(API_URL)
    return res.json()
  }

  _sortAppRecordsByApdex (appRecords) {
    return appRecords.sort(Application.byApdexInDescOrder)
  }

  _createAppModelsFromRecords (appRecords) {
    for (const { host, ...appData } of appRecords) {
      const app = new Application(appData)

      this._dashboard.addApp(app)
      this.addAppToHosts(app, host)
    }
  }

  _getOrCreateHostById (hostId) {
    if (!this._hostIdExists(hostId)) {
      return this._createHostById(hostId)
    }

    const host = this._getHostById(hostId)

    return host
  }

  _getHostById (hostId) {
    return this._dashboard.getHostById(hostId)
  }

  _createHostById (hostId) {
    const host = new Host({ id: hostId })

    this._dashboard.addHost(host)

    return host
  }

  _hostIdExistsOrThrow (hostId) {
    if (!this._hostIdExists(hostId)) {
      throw new Error(
        `The host identified by "${hostId} do not exist."`
      )
    }
  }

  _appExistsOrThrow (app) {
    if (!this._appExists(app)) {
      throw new Error(
        `The app ${app.name} do not exist in the dashboard.`
      )
    }
  }

  _hostIdExists (hostId) {
    return this._dashboard.hasHostId(hostId)
  }

  _appExists (app) {
    return this._dashboard.hasAppId(app.id)
  }
}
