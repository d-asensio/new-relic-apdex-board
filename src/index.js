
import './styles/index.scss'

import { DashboardController } from './controllers'
import { DashboardService } from './services'
import { DashboardView } from './views'

const controller = new DashboardController(
  new DashboardService(),
  new DashboardView()
)

function renderApp () {
  const appRoot = document.getElementById('app')
  controller.renderToDom(appRoot)
}

;(async function IIFE () {
  await controller.init()

  if (document.readyState === 'complete') {
    renderApp()
  } else {
    document.addEventListener(
      'DOMContentLoaded',
      () => renderApp()
    )
  }
})()
