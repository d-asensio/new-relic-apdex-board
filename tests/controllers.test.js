import { enableFetchMocks } from 'jest-fetch-mock'

import { DashboardController } from '../src/controllers'
import { DashboardService } from '../src/services'
import { DashboardView } from '../src/views'

import mockData from '../data/host-app-data.json'

enableFetchMocks()

const smallResponseMock = JSON.stringify(mockData.slice(0, 2))

describe('Dashboard controller', () => {
  beforeEach(() => {
    fetch.resetMocks()
    fetch.mockResponseOnce(smallResponseMock)
  })

  it('is constructed', () => {
    const dashboardController = new DashboardController(
      new DashboardService(),
      new DashboardView()
    )

    expect(dashboardController).toBeInstanceOf(DashboardController)
  })

  it('is initialized', async () => {
    const dashboardController = new DashboardController(
      new DashboardService(),
      new DashboardView()
    )

    await dashboardController.init()

    expect(dashboardController.isInitialized()).toBe(true)
  })

  it('is not initialized', () => {
    const dashboardController = new DashboardController(
      new DashboardService(),
      new DashboardView()
    )

    expect(dashboardController.isInitialized()).toBe(false)
  })

  it('is rendered to a DOM element', async () => {
    const dashboardController = new DashboardController(
      new DashboardService(),
      new DashboardView()
    )

    await dashboardController.init()

    const element = document.createElement('div')

    dashboardController.renderToDom(element)
    const dashboardElement = element.querySelector('.Dashboard')

    expect(dashboardElement).toBeInstanceOf(HTMLElement)
  })

  it('throws an error when attempting to render before initializing', () => {
    const dashboardController = new DashboardController(
      new DashboardService(),
      new DashboardView()
    )

    const element = document.createElement('div')

    expect(() => {
      dashboardController.renderToDom(element)
    }).toThrowErrorMatchingSnapshot()
  })

  it('throws an error when attempting to render to a invalid dom node', async () => {
    const dashboardController = new DashboardController(
      new DashboardService(),
      new DashboardView()
    )

    await dashboardController.init()

    expect(() => {
      dashboardController.renderToDom('.Class')
    }).toThrowErrorMatchingSnapshot()
  })

  it('toggles the layout to list', async () => {
    const dashboardController = new DashboardController(
      new DashboardService(),
      new DashboardView()
    )

    await dashboardController.init()

    const element = document.createElement('div')

    dashboardController.renderToDom(element)

    const toggleElement = element.querySelector('.Toggle')
    toggleElement.click()

    const dashboardElement = element.querySelector('.Dashboard')

    expect(
      dashboardElement.classList.contains('Dashboard--listLayout')
    ).toBe(true)
  })

  it('toggles the layout back to grid', async () => {
    const dashboardController = new DashboardController(
      new DashboardService(),
      new DashboardView()
    )

    await dashboardController.init()

    const element = document.createElement('div')

    dashboardController.renderToDom(element)

    const toggleElement = element.querySelector('.Toggle')
    toggleElement.click()
    toggleElement.click()

    const dashboardElement = element.querySelector('.Dashboard')

    expect(
      dashboardElement.classList.contains('Dashboard--listLayout')
    ).toBe(false)
  })

  it('alerts when clicking an app', async () => {
    const dashboardController = new DashboardController(
      new DashboardService(),
      new DashboardView()
    )

    await dashboardController.init()

    const element = document.createElement('div')

    dashboardController.renderToDom(element)

    const appElement = element.querySelector('.Host__App')

    jest.spyOn(window, 'alert').mockImplementation(alertText => {
      expect(alertText).toMatchSnapshot()
    })

    appElement.click()
  })
})
