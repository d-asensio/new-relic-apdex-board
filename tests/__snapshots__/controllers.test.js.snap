// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard controller alerts when clicking an app 1`] = `"Release number: 7"`;

exports[`Dashboard controller throws an error when attempting to render before initializing 1`] = `"The controller is not yet initialized. You should call to the \\"init ()\\" asynchronous method and await for itto finis before proceeding."`;

exports[`Dashboard controller throws an error when attempting to render to a invalid dom node 1`] = `"The provided element is not valid."`;
